RecoCTPPS/TotemRPLocal/src/FastLineRecognition.cc

```diff
@@ -0,0 +1,308 @@
+    for (vector<Point>::iterator dit = points.begin(); dit != points.end(); ++dit)
+      if (dit->usable)
+        u_points_b++;
+    printf("\tusable points before: %u\n", u_points_b);
+#endif
+
+    // remove points belonging to the recognized line
+    for (vector<const Point *>::iterator hit = c.contents.begin(); hit != c.contents.end(); ++hit)
+    {
+      for (vector<Point>::iterator dit = points.begin(); dit != points.end(); ++dit)
+      {
+        //printf("\t\t1: %.2f, %p vs. 2: %.2f, %p\n", (*hit)->z, (*hit)->hit, dit->z, dit->hit);
+        if ((*hit)->hit == dit->hit)
+        {
+          dit->usable = false;
+          //points.erase(dit);
+          break;
+        }
+      }
+    }
+
+#if CTPPS_DEBUG > 0
+    unsigned int u_points_a = 0;
+    for (vector<Point>::iterator dit = points.begin(); dit != points.end(); ++dit)
+      if (dit->usable)
+        u_points_a++;
+    printf("\tusable points after: %u\n", u_points_a);
+#endif
+  }
+
+#if CTPPS_DEBUG > 0
+  printf("patterns at end: %lu\n", patterns.size());
+  printf("<<<<<<<<<<<<<<<<<<<\n");
+#endif
+}
+
+//----------------------------------------------------------------------------------------------------
+
+bool FastLineRecognition::getOneLine(const vector<FastLineRecognition::Point> &points,
+  double threshold, FastLineRecognition::Cluster &result)
+{
+#if CTPPS_DEBUG > 0
+  printf("\tFastLineRecognition::getOneLine\n");
+#endif
+
+  if (points.size() < 2)
+        return false;
+  
+  vector<Cluster> clusters;
+
+  // go through all the combinations of measured points
+  for (vector<Point>::const_iterator it1 = points.begin(); it1 != points.end(); ++it1)
+  {
+    if (!it1->usable)
+      continue;
+
+    for (vector<Point>::const_iterator it2 = it1; it2 != points.end(); ++it2)
+    {
+      if (!it2->usable)
+        continue;
+
+      const double &z1 = it1->z;
+      const double &z2 = it2->z;
+
+      if (z1 == z2)
+        continue;
+
+      const double &p1 = it1->h;
+      const double &p2 = it2->h;
+      
+      const double &w1 = it1->w;
+      const double &w2 = it2->w;
+
+      // calculate intersection
+      double a = (p2 - p1) / (z2 - z1);
+      double b = p1 - z1 * a;
+      double w = w1 + w2;
+
+#if CTPPS_DEBUG > 0
+      printf("\t\t\tz: 1=%+5.1f, 2=%+5.1f | U/V: 1=%+6.3f, 2=%+6.3f | a=%+6.3f rad, b=%+6.3f mm, w=%.1f\n", z1, z2, p1, p2, a, b, w);
+#endif
+
+      // add it to the appropriate cluster
+      bool newCluster = true;
+      for (unsigned int k = 0; k < clusters.size(); k++)
+      {
+        Cluster &c = clusters[k];
+        if (c.S1 < 1. || c.Sw <= 0.)
+          continue;
+
+#if CTPPS_DEBUG > 0
+        if (k < 10)
+          printf("\t\t\t\ttest cluster %u at a=%+6.3f, b=%+6.3f : %+6.3f, %+6.3f : %i, %i\n", k, c.Saw/c.Sw, c.Sbw/c.Sw, 
+            chw_a, chw_b,
+            (abs(a - c.Saw/c.Sw) < chw_a), (abs(b - c.Sbw/c.Sw) < chw_b));
+#endif
+
+        if ((abs(a - c.Saw/c.Sw) < chw_a) && (abs(b - c.Sbw/c.Sw) < chw_b))
```

Comment:
I can put `std::` to resolve the ambiguity. For my curiosity, what is wrong with `fabs`?